idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,284 @@
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ) {
+        // Using hackyAllowRunningOnEdt here because we don't want to pre-populate all possible textual overrides before user selection.
+        val commands = hackyAllowRunningOnEdt {
+            val batches = analyse(classOrObject) {
+                this.generateMembers(editor, classOrObject, selectedElements, copyDoc)
+            }
+            // Reference shortening is done in a separate analysis session because the session need to be aware of the newly generated
+            // members.
+            analyse(classOrObject) {
+                batches.map { batch ->
+                    val first = batch.first()
+                    val last = batch.last()
+                    collectPossibleReferenceShortenings(first.containingKtFile, TextRange(first.startOffset, last.endOffset))
+                }
+            }
+        }
+        runWriteAction {
+            commands.forEach { it.invokeShortening() }
+        }
+    }
+
+    /**
+     * @returns the generated PSI element inside the target class. The members are grouped together if they are near each other. Each group contains the members in the order they appear in the target class after code generation
+     *
+     */
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun KtAnalysisSession.generateMembers(
+        editor: Editor,
+        currentClass: KtClassOrObject,
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ): List<Iterable<KtDeclaration>> {
+        if (selectedElements.isEmpty()) return emptyList()
+        val selectedMemberSymbols: Map<KtCallableSymbol, KtCallableDeclaration> = buildMap {
+            for (it in selectedElements) {
+                put(it.symbol, generateMember(currentClass.project, it, currentClass, copyDoc))
+            }
+        }
+
+        val classBody = currentClass.body
+        val offset = editor.caretModel.offset
+
+        // Insert members at the cursor position if the cursor is within the class body.
+        if (classBody == null || classBody.textRange.contains(offset)
+            && classBody.lBrace?.textRange?.contains(offset) == false && classBody.rBrace?.textRange?.contains(offset) == false
+        ) {
+            return buildList {
+                insertMembersAfter(
+                    editor,
+                    currentClass,
+                    selectedMemberSymbols.values,
+                    referenceShortener = { collectInsertedBatches(it, this) })
+            }
+        }
+
+        // Insert members at positions such that the result aligns with ordering of members in super types.
+        val symbolsInPreferredOrder = computePreferredOrderOfMembers(currentClass, selectedMemberSymbols)
+        return insertMembersAccordingToPreferredOrder(symbolsInPreferredOrder, classBody.lBrace, editor, currentClass)
+    }
+
+    private fun KtAnalysisSession.computePreferredOrderOfMembers(
```

Comment:
OK, let me add a method doc explaning how it works.

Sorting does not work here because given any two members, we cannot effecitvely determine which member should go before the other by just looking at these two members. Any such decisions must take the order of other members into question.

FE1.0 doesn't have a manually implented doubly linked list because it's using the PSI structure as this doubly linked list. During insetion, it keeps modifying the underlying PSI structure and then does resolve on top of the modified content. If I implement the logic in the exact same way, we will have to create many KtAnalysisSession and rebuild the FIR tree every time we insert a member, that seems to be very inefficient.